contrib/checkstyle/test/src/mill/contrib/checkstyle/CheckstyleModuleTest.scala

```diff
@@ -0,0 +1,154 @@
+import mill.testkit.TestBaseModule
+import utest._
+
+object CheckstyleModuleTest extends TestSuite {
+
+  val resources = os.Path(sys.env("MILL_TEST_RESOURCE_FOLDER"))
+
+  def tests = Tests {
+
+    test("checkstyle") {
+
+      test("format") {
+
+        test("plain") {
+          assert(
+            cs("plain").checkJava(resources / "compatible-java"),
+            cs("plain").checkJava(
+              resources / "sbt-checkstyle",
+              "Utility classes should not have a public or default constructor"
+            ),
+            cs("plain").checkScala(resources / "compatible-scala")
+          )
+        }
+
+        test("sarif") {
+          assert(
+            cs("sarif").checkJava(resources / "compatible-java"),
+            cs("sarif").checkJava(
+              resources / "sbt-checkstyle",
+              "Utility classes should not have a public or default constructor"
+            ),
+            cs("sarif").checkScala(resources / "compatible-scala")
+          )
+        }
+
+        test("xml") {
+          assert(
+            cs("xml").checkJava(resources / "compatible-java"),
+            cs("xml").checkJava(
+              resources / "sbt-checkstyle",
+              "Utility classes should not have a public or default constructor"
+            ),
+            cs("xml").checkJava(
+              resources / "sbt-checkstyle-xslt",
+              "Utility classes should not have a public or default constructor"
+            ),
+            cs("xml").checkScala(resources / "compatible-scala")
+          )
+        }
+      }
+
+      test("options") {
+
+        assert(
+          // Checkstyle exits with org.apache.commons.cli.UnrecognizedOptionException for any option
+          // cs("xml", "6.3", options = Seq("-d")).checkJava(resources / "compatible-java"),
+          cs("xml", options = Seq("-d")).checkJava(resources / "compatible-java")
+        )
+      }
+
+      test("version") {
+
+        test("plain") {
+          assert(
+            // instead of exiting Checkstyle generates a report with a cryptic error
+            cs("plain", "6.2").checkJava(resources / "compatible-java", "File not found"),
+            cs("plain", "6.3").checkJava(resources / "compatible-java")
+          )
+        }
+
+        test("sarif") {
+          intercept[UnsupportedOperationException](
+            cs("sarif", "8.42").checkJava(resources / "compatible-java")
+          )
+          assert(
+            cs("sarif", "8.43").checkJava(resources / "compatible-java")
+          )
+        }
+
+        test("xml") {
+          assert(
+            // instead of exiting Checkstyle generates a report with a cryptic error
+            cs("xml", "6.2").checkJava(resources / "compatible-java", "File not found"),
+            cs("xml", "6.3").checkJava(resources / "compatible-java")
+          )
+        }
+      }
+    }
+  }
+
+  case class cs(format: String, version: String = "10.18.1", options: Seq[String] = Nil) {
```

Comment:
Let's name this `Tester` or something to be a bit less cryptic

Reply:
How about `Checker`?

Reply:
Sure

Reply:
I switched to a more idiomatic approach.